src/main/java/com/google/api/generator/gapic/model/HttpRuleBindings.java

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.gapic.model;
+
+import com.google.api.generator.gapic.utils.JavaStyle;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
+
+@AutoValue
+public abstract class HttpRuleBindings {
+  // private static final Pattern PATH_PARAM_PATTERN = Pattern.compile("[]")
+  public abstract String httpVerb();
+
+  public abstract String pattern();
+
+  public abstract Set<String> pathParameters();
+
+  public abstract Set<String> queryParameters();
+
+  public abstract Set<String> bodyParameters();
+
+  // Private.
+  public static HttpRuleBindings.Builder builder() {
+    return new AutoValue_HttpRuleBindings.Builder()
+        .setHttpVerb("")
+        .setPattern("")
+        .setPathParameters(ImmutableSet.of())
+        .setQueryParameters(ImmutableSet.of())
+        .setBodyParameters(ImmutableSet.of());
```

Comment:
It means that this http binding is empty, representing the case of an rpc in protobuf without `option (google.api.http)` option.

Reply:
In that case, we probably shouldn't be able to instantiate an HttpRuleBindings at all (or we'd be able to set an empty pattern and verb with non-empty parameter lists). Could we do the following instead?

- Make `HttpRuleBindings` `Nullable` on `Method`
- Remove the empty string setters here, and add checks in `build()` to fail on invalid cases (i.e. empty string, pattern has no slashes and is not PubSub's `_deleted-topic_`, etc).
- Consider making the verb an enum, or check that the string is one of the accepted five values in `build()`. (Would prefer the enum, since the latter is more error-prone.)

Context: Using rigorous validity checks has really helped me with development in this codebase. For instance, adding type-checking on assignment expressions or requiring set types on methods has helped me catch those errors at compile-time, so I think more rigid checks would help us here as well.

Reply:
Please check the other comment about `HttpRuleBindings` for completeness. In addition to that:
- In current design `HttpRuleBindings` has a notion of being empty, which corresponds to the case when there is no `google.api.http` option at all for a method. This was intentional. Converting it to `nullable` will change the semantics to represent the "no  `google.api.http`" casse as `null ` for `HttpRuleBindings`. Both will work, but overall it is usually a better practice to not do `null`s if possible, because they are very non-specific, non-static (nulls of every type are equal) and error-prone (for example, does `null` mean that there was no bindings, or that the method hasn't been initialized)
- By the current design, `HttpRuleBindings` is a POJO basically, which does not even know that it was populated from `google.api.http` (and it does not have to be, for mixins, or service.yaml, where we still may have http bindings, but initialized from yaml instead of `google.api.http`). That is why this class is not responsible for validating any `gogole.api.http` specific stuff.
- `HttpRuleParser` - this is the class which is supposed to be doing parsing and validation before building an instance of this class, and it already does it (we can add more there, if necessary).

I guess, I named the class poorly. It shoudl have been named HttpBindings, to get rid of the "Rule" name in it to clear the confusion and make it clear that `HttpBindings` hold the bindings without any specific information where those bidings came from (from google.api.http or from service yaml or just put manually).

Reply:
Appending from our discussion offline: Deferring the pattern check to a future PR is fine.